### ISSUE TITLE
Support for TYPO3 9.x

### ISSUE THREAD

**Issue by fnagel** [fnagel/t3extblog#185]
**Work in progress branch:**

~~https://github.com/fnagel/t3extblog/tree/typo3-9~~
~~https://github.com/fnagel/t3extblog/archive/typo3-9.zip~~

Merged to master branch!

**What needs to be done:**

* [x] Fix ViewHelpers
   * [x]  Adjust for TYPO3 9.x
   * [x] Fix comment allowed VH (has some caching issues)
* [x] Replace TYPO3_DB globals
* [x] Replace EXT:dd_googlesitemap with core feature
* [x] Replace EXT:realurl config with core feature
   * [x] General implementation
      * [x] Add URL segment field for Post
      * [x] Add URL segment field for Category
      * [x] Fix chash parameter issue
   * [x] Fix minor paginator issues
     * [x] Fix parameter for first page (when on e.g. on page 2)
     * [x] ~~Fix that list and page 1 is the same (duplicate content)~~
   * [x] Add update wizard for URL segment fields (see here for example: https://github.com/georgringer/news/blob/master/Classes/Updates/NewsSlugUpdater.php)
   * [x] Blogsystemplugin  
      * [x] General 
      * [x] Improve post date segment (see here for example: https://github.com/jweiland-net/events2/blob/master/Classes/Routing/Aspect/TimestampMapper.php) 
      * [x] Fix tags parameter (create an extra ´Mapper´ for CSV fields)
   * [x] SubscriptionManager plugin
   * [x] BlogSubscription plugin
* [x] Use core SEO features for all meta tag related stuff
* [x] Test BE module
* [x] Test email sending from FE context
  * [x] Single language
  * [x] Multi language
* [x] Test email sending from BE context
  * [x] Single language
  * [x] Multi language
* [x] Fix "Scan Extension Files" BE module issues
   * [x] Fix at least all "strong" issues
   * [x] Remove removeXSS method
* [x] Fix tests
* [x] Documentation
  * [x] Finalize upgrade guide
  * [x] Remove EXT:realurl documentation
  * [x] Add section about new routing
  * [x] Add section about new sitemap

**Comment by fnagel** [comment]
WIP branch: https://github.com/fnagel/t3extblog/tree/typo3-9

**Comment by fnagel** [comment]
@netthinks Nothing special here. Just include the `Default.yml` file (see here: https://github.com/fnagel/t3extblog/blob/typo3-9/Configuration/Routes/Default.yaml) into your site config. 

This might help: 
https://docs.typo3.org/m/typo3/reference-coreapi/master/en-us/ApiOverview/SiteHandling/Basics.html#site-configuration-storage and https://docs.typo3.org/m/typo3/reference-coreapi/master/en-us/ApiOverview/Routing/Examples.html?highlight=imports#usage-with-imports

**Comment by netthinks** [comment]
hmm, tried it but now an error is shown:

(1/1) Symfony\Component\Routing\Exception\InvalidParameterException

Parameter "te9381b972eb5c6fe1ed89e10c57e700" for route "tx_t3extblog_blogsystem_0" must match "[^/]++" ("" given) to generate a corresponding URL.

**Comment by fnagel** [comment]
I did not run into this error so I'm not quite sure what's happening here. Did you try to debug in order to see what part of the route config does cause the error?

**Comment by fnagel** [comment]
@criwe @netthinks @hweckhardt Found some time to work a little more on TYPO3 9.5 support. Almost finished (missing docs and tests need to be fixed, but no functionality missing)! 

I could need some help testing. Anyone in?

Make sure to check out the upgrade guide: https://github.com/fnagel/t3extblog/blob/typo3-9/Documentation/UpgradeGuide/Version-5.rst

**Comment by netthinks** [comment]
used the update script to fill the url_segment, but now all blog articles have the same content "ein-neuer-blogartikel"...
the error on blog page list view is still there:
```
 Symfony\Component\Routing\Exception\InvalidParameterException
Parameter "te9381b972eb5c6fe1ed89e10c57e700" for route "tx_t3extblog_blogsystem_0" must match "[^/]++" ("" given) to generate a corresponding URL.
```

**Comment by fnagel** [comment]
@netthinks 

> used the update script to fill the url_segment, but now all blog articles have the same content "ein-neuer-blogartikel"...

My bad. Fixed in the latest commit!


> the error on blog page list view is still there

Still not able to reproduce. Do I understand correctly that you are not able to view the post list at all? Is it an existing installation you are trying to update? Are all posts valid? Would you mind removing parts of the configuration in `EXT:t3extblog/Configuration/Routes/Blogsystem.yaml` in order to see which one breaks the routing?

**Comment by netthinks** [comment]
I commented out lines 47 through 50, then it worked:
     - routePath: '/{category_prefix_localized}/{category_title}'
        _controller: 'Post::category'
        _arguments:
          'category_title': 'category'

**Comment by fnagel** [comment]
Ok, seems something is wrong with your categories. Do you updated an existing blog? I've just noticed that an update wizard for adding slugs to existing categories is missing. I will create one asap.

Please check if your categories have a valid `url_segment` field.

**Comment by fnagel** [comment]
See this commit for the update script for categories: 5e4bdcb0e6755b4fd8f68ae55f9b94f3fdcb2063

**Comment by fnagel** [comment]
@criwe @netthinks @hweckhardt 

I think the extension is now fully ready for TYPO3 9.5! I could need some final feedback before the release. Anybody able to to some testing? Any help is highly appreciated!

**Comment by netthinks** [comment]
> 
> 
> Ok, seems something is wrong with your categories. Do you updated an existing blog? I've just noticed that an update wizard for adding slugs to existing categories is missing. I will create one asap.
> 
> Please check if your categories have a valid `url_segment` field.

Categories were empty, with the new update script everything worked fine

**Comment by fnagel** [comment]
Nice. That was an easy fix :-D

**Comment by fnagel** [comment]
@criwe @netthinks @hweckhardt 

Sorry to annoy you again, but I could really need some more feedback for the finale release. Anybody able to help?

FYI: I've just merged the `typo3-9` branch into `master`!

**Comment by netthinks** [comment]
works very well so far, but I've found one problem: after releasing a comment, an error occurs. Don't know if it's a problem of my config, can you check please? The comment is released after reload...
![02-11-_2019_14-15-00](https://user-images.githubusercontent.com/23073131/68071510-69d6cb80-fd7b-11e9-9cae-39e979945638.png)
![02-11-_2019_14-15-42](https://user-images.githubusercontent.com/23073131/68071511-6a6f6200-fd7b-11e9-93b5-9a6d9b2038cf.png)

**Comment by fnagel** [comment]
@netthinks Please give me some more insight on what exactly happened. I've tried the following:

* Go to an article
* Add comment (no spam, but needs approval)
* Go to BE module
* Confirm comment by clicking the little red x icon
* Reload article page

How to reproduce the issue in your case?

In addition: please add the complete stack trace!

**Comment by netthinks** [comment]
@fnagel 
- Go to an article
- Add comment (no spam, but needs approval)
- Go to BE module
- Confirm comment by clicking the little red x icon: error occurs

Stack trace: https://pastebin.com/HESjkq8J

**Comment by fnagel** [comment]
@netthinks Mhhh, hard to say what's the issue here. Looks like the link VH within the email templates crashes. Is the page UID 244 (should be page with blogsystem plugin) and the post UID 33 correct? 

Can you post your TS settings?
Are you using default templates? 
How do you add the TS (additional template or in the root template)?

**Comment by netthinks** [comment]
Tested, now runs without problems. Thank you!

**Comment by fnagel** [comment]
Closed with a019d1ccea4f0f98f18f3ea65f44b5ec3677c90c
